File: homing-core-js/src/main/java/hue/captains/singapura/js/homing/core/js/ServingContextModule.java
package hue.captains.singapura.js.homing.core.js;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The query this module was served with — {@code theme} and {@code locale} —
 * read from {@code import.meta.url}, and a helper that carries it onto a bare
 * module URL.
 *
 * <p><b>Why it exists.</b> A module's imports inherit the query its own request
 * carried, so everything under the chrome is keyed by
 * {@code …&theme=T&locale=L}. A widget's {@code moduleUrl} is minted bare, and
 * a bare request starts a <i>second</i> import chain: a second instance of
 * every module the chrome already loaded. For a stateless module that is
 * duplicated code; for {@code DomOpsPartyModule} it is a second, empty root.
 * RFC 0063 found this by building the party monitor and watching it draw a
 * tree with nothing in it.</p>
 *
 * <p><b>Why it is alone.</b> {@code import.meta} does not parse in a classic
 * script, which is how the test harness loads raw module files. Isolating the
 * one read here keeps every consumer testable; a consumer stubs
 * {@code withServingContext} the way it stubs its importer.</p>
 *
 * @since RFC 0063
 */
public record ServingContextModule() implements DomModule<ServingContextModule> {

    public static final ServingContextModule INSTANCE = new ServingContextModule();

    /** {@code servingContext() → { theme, locale }}. */
    public record servingContext() implements Exportable._Constant<ServingContextModule> {}

    /** {@code withServingContext(url) → url} with the context appended where absent. */
    public record withServingContext() implements Exportable._Constant<ServingContextModule> {}

    @Override
    public ImportsFor<ServingContextModule> imports() {
        return ImportsFor.<ServingContextModule>builder().build();
    }

    @Override
    public ExportsOf<ServingContextModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new servingContext(), new withServingContext()));
    }
}

File: homing-core-js/src/main/resources/homing/js/hue/captains/singapura/js/homing/core/js/ServingContextModule.js
// =============================================================================
// ServingContextModule — the query this module was served with.
//
//   servingContext()      → { theme: string|null, locale: string|null }
//   withServingContext(u) → u, with &theme= and &locale= appended if absent
//
// A module's imports inherit the query its OWN request carried: the chrome is
// served as /module?class=…&theme=T&locale=L, and every import it emits carries
// the same, so the browser keys every module under it by that URL. A request
// that starts BARE starts a second chain — a second instance of every module
// the first chain already loaded. Harmless for a stateless module; for one
// that holds a singleton it is a split brain. RFC 0063 found it by building a
// monitor for the party tree and having it draw an empty second root.
//
// import.meta.url is the one honest source: it is the URL THIS module was
// loaded from, and since this module is imported by the chrome's chain it
// carries the chrome's query. It lives alone here because import.meta does not
// parse in a classic script, which is how the test harness loads raw modules;
// anything that needs the context imports this and stubs it under test.
// =============================================================================

function servingContext() {
    try {
        const p = new URL(import.meta.url).searchParams;
        return { theme: p.get('theme'), locale: p.get('locale') };
    } catch (e) {
        return { theme: null, locale: null };
    }
}

function withServingContext(url) {
    const ctx = servingContext();
    let out = String(url);
    if (ctx.theme  && !/[?&]theme=/.test(out))  out += '&theme='  + encodeURIComponent(ctx.theme);
    if (ctx.locale && !/[?&]locale=/.test(out)) out += '&locale=' + encodeURIComponent(ctx.locale);
    return out;
}
